### PR TITLE
Adding the cargo doc --examples subcommand

### PR DESCRIPTION
**Pull request by QiangHeisenberg**
Adding the cargo doc --examples subcommand

#4508

### CODE REVIEW

File: src/bin/cargo/commands/doc.rs [hunk 1]
```diff
@@ -12,6 +12,7 @@ pub fn cli() -> App {
             "open",
             "Opens the docs in a browser after the operation",
         ))
+        .arg(opt("examples", "Document all examples"))
```
Comment (jyn514) [inline src/bin/cargo/commands/doc.rs:15]: This doesn't actually do anything; if someone passes the option cargo will ignore it. Consider writing some tests so you can tell whether this works or not, there are lots of examples if you search for  `cargo_test`.

I'm not super familiar with cargo's code, but I expect the right approach would be to loop over all examples and document them (maybe in parallel?). I'd look at how --example x works currently.

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/bin/cargo/commands/doc.rs (hunk 1 above)
Comment (QiangHeisenberg) [reply src/bin/cargo/commands/doc.rs:15]: Thank you for your review, which is a huge encouragement to me. It works fine for my local test.

My test file directory is:
```
hello
├─examples
│    ├───examples1.rs
│    └───examples2.rs
└─src
   └───main.rs
```
image

I think it's good extensibility in the source code.
```
..\cargo\src\cargo\util\command_prelude.rs
```
```
filter: CompileFilter::from_raw_arguments(
                self._is_present("lib"),
                self._values_of("bin"),
                self._is_present("bins"),
                self._values_of("test"),
                self._is_present("tests"),
                self._values_of("example"),
                self._is_present("examples"),     // examples
                self._values_of("bench"),
                self._is_present("benches"),
                self._is_present("all-targets"),
            ),
```

**Comment by ehuss** [comment]
Thanks for the PR!

Just a few questions/comments:

* Can you say more about the use case here? Do you have a public project that demonstrates where this would be useful?  It seems a bit unusual since you can't use example libraries as a dependency.
* I would probably just update `arg_targets_lib_bin` to add examples instead of adding a separate argument here, since `cargo doc` is the only command that uses `arg_targets_lib_bin`.
* This should probably support both `--examples` and `--example`.
* All additions need tests to be added. There's more information at https://doc.crates.io/contrib/tests/index.html about writing and running tests.
* CLI changes need to be added to the man pages. There is some documentation [here](https://github.com/rust-lang/cargo/tree/master/src/doc#building-the-man-pages) that explains how that works.

File: src/bin/cargo/commands/doc.rs [hunk 2]
```diff
@@ -21,10 +21,12 @@ pub fn cli() -> App {
         .arg(opt("no-deps", "Don't build documentation for dependencies"))
         .arg(opt("document-private-items", "Document private items"))
         .arg_jobs()
-        .arg_targets_lib_bin(
+        .arg_targets_lib_bin_example(
             "Document only this package's library",
             "Document only the specified binary",
             "Document all binaries",
+            "Document only the specified binary",
```
Comment (jyn514) [inline src/bin/cargo/commands/doc.rs:28]:
```suggestion
            "Document only the specified example",
```

Comment (QiangHeisenberg) [reply src/bin/cargo/commands/doc.rs]: thanks.done.

**Comment by QiangHeisenberg** [comment]
Frankly, I don't have any more use cases here.I think it is a legacy issue.Maybe you're busy, so I'll help you out.

In addition, the test items I conducted locally were normal, but I failed here. Why?

**Comment by jyn514** [comment]
> Can you say more about the use case here? Do you have a public project that demonstrates where this would be useful? It seems a bit unusual since you can't use example libraries as a dependency.

For consistency with the other cargo subcommands. It seems weird you can use `cargo check --examples` and `cargo doc --example x` but not `cargo doc --examples`.

(That said, I agree I'm not sure how useful documenting examples is, I think https://github.com/rust-lang/rfcs/pull/3123 is more useful.)

**Comment by ehuss** [comment]
> In addition, the test items I conducted locally were normal, but I failed here. Why?

When changing the man pages, the `build-man.sh` script needs to be run to rebuild them, and then you need to check in all those changes.

**Comment by QiangHeisenberg** [comment]
> > In addition, the test items I conducted locally were normal, but I failed here. Why?
> 
> When changing the man pages, the `build-man.sh` script needs to be run to rebuild them, and then you need to check in all those changes.

thanks for your advice. It's normal now.

**Review by jyn514 — APPROVED** [review]

**Comment by ehuss** [comment]
Thanks!

@bors r+

**Pull request merged**